Export PureI18nProvider without side-effects for testing purposes

FredyC opened the pull request:
I know, it's uncommon to have a whole Context exported but bear with me.

Current `I18nProvider` includes a side effect to provide an update to context when `i18n` changes something. That's cool and even though I know I wouldn't use that, it's not the main reason for this PR.

https://github.com/lingui/js-lingui/blob/f6167e0cff98de6ffd7da2296058b0405af5fbeb/packages/react/src/I18nProvider.tsx#L36-L39

When running tests, this side effect will cause an infamous `act warning` and I haven't found a way on how to prevent that except removing that side effect from the code.

![image](https://user-images.githubusercontent.com/1096340/82113315-68885180-9755-11ea-87f7-c9ab57469af0.png)

With exported Context and I can have my own Provider in tests without that side effect.

An alternative solution would be to introduce a boolean to props that would disable that side effect, but that doesn't sound great.

PR is a draft, for now, I will update documentation when we have agreed on a way to go.

**Bonus**

I have exported `usei18n` hook purely as a shortcut because in most cases users don't care about `defaultRender`.

cc @tricoder42

FredyC commented on packages/react/src/I18nProvider.tsx:
```diff
@@ -1,15 +1,15 @@
 }
 
-export interface I18nProviderProps extends I18nContext {}
+export type I18nProviderProps = I18nContext
```
Just stumbled upon a perfect article for you today if you think you cannot extend type :)

https://ultimatecourses.com/blog/typescript-interfaces-vs-types

Ultimately, in most cases, it does boil down to personal preference. For me it's based on the fact that `type` can do pretty much everything (and more) compared to `interface`. Besides there is also `import type` since 3.8, no `import interface`. Using two different keywords becomes opinionated battle when use which. So I opted to `type` everywhere and it's a win.

FredyC commented on packages/react/src/I18nProvider.tsx:
```diff
@@ -23,6 +23,10 @@ export function useLingui(): I18nContext {
   return context
 }
 
+export function usei18n(): I18n {
+  return useLingui().i18n
+}
+
```
And why would everyone need to implement it when you can export here? How does it hurt anyone? It's ESM, with proper tree shaking setup it won't even be in the bundle if people don't use it. It's needed rarely of course, but still, I don't see a reason why not to have it.

FredyC commented:
> Hi @FredyC and thanks for this PR. I agree that we should export context if it's helpful, but maybe the problem is deeper. I guess even if you implement your own provider, other users will end up in the same position as you do. What if we figure out how to approach testing and document this instead? If we require a different provider for testing, maybe we could make it part of the library?

Interesting, so you don't want to include purely functional `useI18n` but you are ok with including something for testing purposes only? 😆 I am fine with it, just find such logic kinda funny.

tricoder42 commented:
> Interesting, so you don't want to include purely functional useI18n but you are ok with including something for testing purposes only? 😆 I am fine with it, just find such logic kinda funny.

Well, the reason is that `usei18n` doesn't add a new value, while a testing provider does add a new value. `usei18n` doesn't include any additional logic, while testing provider might require knowledge of library internals. Libraries usually expose testing utilities which makes it easier.

> My point is, as a consumer, you don't care about defaultRender prop, that's an internal thing used by other parts. If you want to get access to i18n, why you should go through extra hoops?

That's my point too. If we don't need `defaultRender` prop most of the time and it definitely doesn't depend on `i18n` object, we could also split context into two.

> https://ultimatecourses.com/blog/typescript-interfaces-vs-types

I'll check the article later today, but if we want to start using `types` instead of `interface`, then we should migrate the rest of the codebase to make it consistent.

FredyC commented:
Made PureI18nProvider (better name suggestions?), removed useI18n, and export of context.



> I'll check the article later today, but if we want to start using `types` instead of `interface`, then we should migrate the rest of the codebase to make it consistent.

I am fine working on a separate PR to migrate it if you don't have some extra love for `interface` :)

tricoder42 commented:
> I am fine working on a separate PR to migrate it if you don't have some extra love for interface :)
Merge state

I don't care at all. Whatever works. Thanks for working on this 👍

tricoder42 commented:
Looks good. Let's try it until we find a better way to support testing.

Could youp please also add it to [reference docs](https://github.com/lingui/js-lingui/blob/next/docs/ref/react.rst). Bonus points if you also wanted to write a short [testing guide](https://github.com/lingui/js-lingui/blob/next/docs/guides/testing.rst) (the original one can be removed completely)

FredyC commented:
Sorry for the delay here. I am not entirely sure what would I put into docs. My testing strategy consists of using source locale to have a predictable output. And if a need to test actual localization arises, the E2E is probably a much better choice.

Anyway, I am even more convinced now that the side effect of listening to `change` event should be at least opt-out even for production code. I noticed it's a source of whole bunch of extra re-renders when activating locale upon the app start. I want to avoid that.

FredyC commented:
So yea, I dropped `PureI18nProvider` again and added a boolean flag `optOutFromChanges` to `i18nProvider` along with an update to ref docs. I believe this is a better approach.

tricoder42 commented:
Looks good, thanks 👍  I'm just thinking about a better name than `optOutFromChanges` 🤔

FredyC commented:
@tricoder42 I think this is still very much relevant. Your testing docs are too simplified and I am pretty sure the `act warning` will still happen in more complex cases.

Sadly, I don't really have the capacity right now to handle this, but I am pretty sure that adding boolean prop and opt-out behavior from that effect is pretty easy for anyone around here.
